src/mlpack/methods/decision_tree/mse_gain.hpp

```diff
@@ -96,54 +96,179 @@ class MSEGain
   }
 
   /**
-   * Calculates the weighted mean squared error gain given the sum of squares
-   * and mean.
+   * Calculates the  mean squared error gain for the given child and index.
    *
    * X = array of values of size n.
-   * W = array of weights of size n.
    *
    * @f{eqnarray*}{
-   *    MSE = \sum\limits_{i=1}^n {W_i * {X_i}^2} -
-   *        {\dfrac{\sum\limits_{j=1}^n W_j * X_j}
-   *        {\sum\limits_{j=1}^n W_i}}^2
+   *   MSE = \sum\limits_{i=1}^n {X_i}^2 -
+   *       {\dfrac{\sum\limits_{j=1}^n X_j}{n}}^2
    * @f}
    *
-   * @param weightedSumSquares Precomputed weighted sum of square
-   *    (sum(Wi * Xi^2)) of values.
-   * @param weightedMean Precomputed weighted mean (sum(Wi * Xi) / sum(Wi)) of
-   *    values.
-   * @param totalChildWeight Total weight of all the samples in that child.
+   * @param index The current index to calculate gain.
+   * @param child The child to calculate gain.
+   *    0 -> Left child,   1 -> Right child
    */
-  static double Evaluate(const double weightedSumSquares,
-                         const double weightedMean,
-                         const double totalChildWeight)
+  double Evaluate(const size_t index, const size_t child)
   {
-    double mse = weightedSumSquares / totalChildWeight -
-        weightedMean * weightedMean;
+    double mse;
+    // Left child.
+    if (child == 0)
+      mse = sumSquares[index] / leftSize - leftMean * leftMean;
+    // Right child.
+    else
+      mse = (totalSumSquares - sumSquares[index]) / rightSize
+          - rightMean * rightMean;
     return -mse;
   }
 
   /**
-   * Calculates the  mean squared error gain given the sum of squares and mean.
+   * Caches the prefix sum of squares to efficiently compute gain value for
+   * each split. It also computes the initial mean for left and right child.
    *
-   * X = array of values of size n.
-   *
-   * @f{eqnarray*}{
-   *   MSE = \sum\limits_{i=1}^n {X_i}^2 -
-   *       {\dfrac{\sum\limits_{j=1}^n X_j}{n}}^2
-   * @f}
+   * @param responses The set of responses on which statistics are computed.
+   * @param weights The set of weights associated to each response.
+   * @param minimum The minimum number of elements in a leaf.
+   */
+  template<bool UseWeights, typename ResponsesType, typename WeightVecType>
+  void CalculateStatistics(const ResponsesType& responses,
```

Comment:
Take a look at https://github.com/RishabhGarg108/mlpack-1/blob/ad64ad07d717c2b2d19be7bac82be03a2329071b/src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp#L407 from lines 407 to 417. First, we update statistics, then we do a check if we can skip the gain computation for a particular index and then we evaluate gain and later do other stuff.

Now, if we want to combine the `UpdateStatistics` and `Evaluate` method, then we have to move this check that skips the gain computation into that function too and it will be ugly to `continue` the loop from the function. We will have to create a flag variable that will be set to true when the skip condition is true in the function and based on that we will have to put another condition in the loop to `continue` it.

So, I think we can keep them separated. Let me know if I overlooked something or if there is some other way to achieve this.

One thing that we can definitely do is to return a tuple from the `Evaluate` method. 👍 

Also, I am not that good at naming functions. Can you tell the exact names for these functions that you would like? :)

Reply:
Ahh, so the only thing I was thinking is that the update on line 407 can be removed, but then you would need to have the function that computes the gain allow updating the index by many points at once (like described [here](https://github.com/mlpack/mlpack/pull/2905#discussion_r664930851)).  But I do think either way is fine, so, up to you.

For names I might suggest:

 - `BinaryScanInitialize()`
 - `BinaryStep()` for `UpdateStatistics()` (since it 'step's one index at a time)
 - `BinaryGains()` to compute the left and right gains

Let me know what you think. :+1:

Reply:
The names make sense. Thanks :)